src/allmydata/test/cli/test_daemonize.py

```diff
@@ -0,0 +1,104 @@
+from mock import patch, Mock
+from StringIO import StringIO
+from sys import getfilesystemencoding
+from twisted.trial import unittest
+from allmydata.scripts import runner
+from allmydata.scripts.tahoe_daemonize import identify_node_type
+from allmydata.scripts.tahoe_daemonize import DaemonizeTahoeNodePlugin
+from allmydata.scripts.tahoe_daemonize import DaemonizeOptions
+
+
+class Util(unittest.TestCase):
+
+    def test_node_type_nothing(self):
+        tmpdir = self.mktemp()
+        base = dirname(tmpdir).decode(getfilesystemencoding())
+
+        t = identify_node_type(base)
+
+        self.assertIs(None, t)
+
+    def test_node_type_introducer(self):
+        tmpdir = self.mktemp()
+        base = dirname(tmpdir).decode(getfilesystemencoding())
+        with open(join(dirname(tmpdir), 'introducer.tac'), 'w') as f:
+            f.write("test placeholder")
+
+        t = identify_node_type(base)
+
+        self.assertEqual(u"introducer", t)
+
+    def test_daemonize(self):
+        tmpdir = self.mktemp()
+        plug = DaemonizeTahoeNodePlugin('client', tmpdir)
+
+        with patch('twisted.internet.reactor') as r:
+            def call(fn, *args, **kw):
+                fn()
+            r.callWhenRunning = call
+            service = plug.makeService(None)
+            service.parent = Mock()
+            service.startService()
+
+        self.assertTrue(service is not None)
+
+    def test_daemonize_no_keygen(self):
+        tmpdir = self.mktemp()
+        plug = DaemonizeTahoeNodePlugin('key-generator', tmpdir)
+
+        with patch('twisted.internet.reactor') as r:
+            def call(fn, *args, **kw):
+                fn()
+            r.callWhenRunning = call
+            service = plug.makeService(None)
+            service.parent = Mock()
+            with self.assertRaises(ValueError) as ctx:
+                service.startService()
+            self.assertIn(
+                "key-generator support removed",
+                str(ctx.exception)
+            )
+
+    def test_daemonize_unknown_nodetype(self):
+        tmpdir = self.mktemp()
+        plug = DaemonizeTahoeNodePlugin('an-unknown-service', tmpdir)
+
+        with patch('twisted.internet.reactor') as r:
+            def call(fn, *args, **kw):
+                fn()
+            r.callWhenRunning = call
+            service = plug.makeService(None)
+            service.parent = Mock()
+            with self.assertRaises(ValueError) as ctx:
+                service.startService()
+            self.assertIn(
+                "unknown nodetype",
+                str(ctx.exception)
+            )
+
+    def test_daemonize_options(self):
+        parent = runner.Options()
+        opts = DaemonizeOptions()
+        opts.parent = parent
+        opts.parseArgs()
+
+        # just gratuitous coverage, ensureing we don't blow up on
+        # these methods.
+        opts.getSynopsis()
+        opts.getUsage()
+
+    def test_daemonize_defaults(self):
```

Comment:
It looks like the `runner.dispatch()` is allowing this to run too far, and it looks at `~/.tahoe` (which isn't a real node directory on most systems, so the command bails with "... doesn't look like a directory at all").

I'd start by adding something to the assertion to make this easier to spot:

```python
    self.assertEqual(0, exit_code[0],
                     [exit_code[0], o.getvalue(), e.getvalue()])
```

and then I think we want two separate steps. The first should build a config from the command with no arguments, and it should look at the resulting config and check that the nodedir is pointing at `~/.tahoe` (or whatever the platform-specific value is.. there's a function in scripts/ somewhere that returns that value). Then we should modify the config object to point at a tempdir, and then allow the second part to run (`runner.dispatch`), and make sure it behaves sensibly.

Or, maybe we should mock out that what-is-the-default-nodedir function, and have it return the tempdir instead?

(what's this testing, exactly? is it that we can run `tahoe daemonize` on a plausibly-valid nodedir without complaint? Are you patching `twistd` to catch it at the last moment before it really spawns the daemon?)